### PR TITLE
[RUN-3526] Cleanup XPath.java and add more functions

### PR DESCRIPTION
**Pull request by atennapel**
This PR:
- Cleans up XPath.java
- Adds `startsWith`, `endsWith`, `gte`, `lt` and `lte`
- Fixes `gt`, which was using `>=` instead of the correct `>`
- Update Guava to 32.0.1

### CODE REVIEW

File: src/test/communitycommons/XPathTest.java [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+    var xpath = XPath.create(null, "A");
+    assertEquals("//A", xpath.getXPath());
+  }
+
+  // comparisons
+  @Test
+  public void eq() throws Exception {
+    var xpath = XPath.create(null, "A").eq("attr", 42);
+    assertEquals("//A[attr = 42]", xpath.getXPath());
+    var xpathPath = XPath.create(null, "A").eq("attr1", "attr2", "attr3", 42);
```
Comment (lvanengelen) [inline src/test/communitycommons/XPathTest.java:56]: 🎄 strictly `attr1` and `attr2` here are association and entity type respectively. Could we name them accordingly here and in later cases?

File: src/test/communitycommons/XPathTest.java [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+  // XPath
+  @Test
+  public void simple() throws Exception {
+    var xpath = XPath.create(null, "A");
+    assertEquals("//A", xpath.getXPath());
+  }
+
+  // comparisons
+  @Test
+  public void eq() throws Exception {
```
Comment (lvanengelen) [inline src/test/communitycommons/XPathTest.java:53]: We should add tests for the assertions on the number of optional parameters passed.

**Review by lvanengelen — CHANGES_REQUESTED** [review]

File: build.gradle [hunk 3]
```diff
@@ -78,6 +78,7 @@ sourceSets {
 
 test {
     exclude 'unittesting/*'
+    testLogging.showStandardStreams true
```
[the remark this answers is not in the document]
Comment (atennapel) [reply build.gradle:81]: Ah I was testing things in the test using println, this is required for the println to show up. I guess we can remove it again.

File: src/CommunityCommons/javasource/communitycommons/XPath.java [hunk 4]
```diff
@@ -33,806 +33,829 @@
...
-	// subset of characters.
-	private static final CharSequenceTranslator ESCAPE_XML =
-		new AggregateTranslator(
-			new LookupTranslator(EntityArrays.BASIC_ESCAPE),
-			new LookupTranslator(EntityArrays.APOS_ESCAPE)
-		);
+public class XPath<T> {
+  /**
+   * Build in tokens, see:
+   * https://world.mendix.com/display/refguide3/XPath+Keywords+and+System+Variables
```
[the remark this answers is not in the document]
Comment (atennapel) [reply src/CommunityCommons/javasource/communitycommons/XPath.java:39]: I think we can remove it then.

File: src/CommunityCommons/javasource/communitycommons/XPath.java [hunk 5]
```diff
@@ -33,806 +33,829 @@
...
-	// Implement our own ESCAPE_XML because the original got deprecated and we originally translated only a very small
-	// subset of characters.
-	private static final CharSequenceTranslator ESCAPE_XML =
-		new AggregateTranslator(
-			new LookupTranslator(EntityArrays.BASIC_ESCAPE),
-			new LookupTranslator(EntityArrays.APOS_ESCAPE)
-		);
+public class XPath<T> {
+  /**
+   * Build in tokens, see:
```
[the remark this answers is not in the document]
Comment (atennapel) [reply src/CommunityCommons/javasource/communitycommons/XPath.java:38]: Oh I didn't think of that. Yes then we should keep that convention.

File: src/CommunityCommons/javasource/communitycommons/XPath.java [hunk 6]
```diff
@@ -33,806 +33,829 @@
...
+  }
+
+  public XPath<T> lte(Object... pathAndValue) {
+    return compare("<=", pathAndValue);
+  }
+
+  public XPath<T> contains(Object attr, String value) {
+    autoInsertAnd().append(" contains(").append(String.valueOf(attr)).append(",").append(valueToXPathValue(value))
+        .append(") ");
+    return this.requireBinOp(true);
```
[the remark this answers is not in the document]
Comment (atennapel) [reply src/CommunityCommons/javasource/communitycommons/XPath.java:200]: Yes good.

File: src/CommunityCommons/javasource/communitycommons/XPath.java [hunk 7]
```diff
@@ -33,806 +33,829 @@
...
+  public XPath<T> lte(Object attr, Object valuecomparison) {
+    return compare(attr, "<=", valuecomparison);
+  }
+
+  public XPath<T> lte(Object... pathAndValue) {
+    return compare("<=", pathAndValue);
+  }
+
+  public XPath<T> contains(Object attr, String value) {
+    autoInsertAnd().append(" contains(").append(String.valueOf(attr)).append(",").append(valueToXPathValue(value))
```
[the remark this answers is not in the document]
Comment (atennapel) [reply src/CommunityCommons/javasource/communitycommons/XPath.java:198]: Yes that would be nice.

File: src/CommunityCommons/javasource/communitycommons/XPath.java [hunk 8]
```diff
@@ -33,806 +33,829 @@
...
+    autoInsertAnd().append(" ends-with(").append(String.valueOf(attr)).append(",").append(valueToXPathValue(value))
+        .append(") ");
+    return this.requireBinOp(true);
+  }
+
+  public XPath<T> compare(Object attr, String operator, Object value) {
+    return compare(new Object[] { attr }, operator, value);
+  }
+
+  public XPath<T> compare(String operator, Object[] pathAndValue) {
```
[the remark this answers is not in the document]
Comment (atennapel) [reply src/CommunityCommons/javasource/communitycommons/XPath.java:219]: Yeah I guess so.

File: src/test/communitycommons/XPathTest.java (hunk 1 above)
Comment (atennapel) [reply src/test/communitycommons/XPathTest.java:56]: Yes that would be better.

File: src/CommunityCommons/javasource/communitycommons/XPath.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (reinouts) [reply src/CommunityCommons/javasource/communitycommons/XPath.java:39]: Before removing it, can I suggest including all the tokens listed on that documentation page? It currently lists 40 tokens if I counted right, this Java code only has 19.

File: src/CommunityCommons/javasource/communitycommons/XPath.java [hunk 9]
```diff
@@ -33,806 +33,829 @@
...
+  public static final String ID = "id";
+  /** End builtin tokens */
+
+  private String entity;
+  private int offset = 0;
+  private int limit = -1;
+  // important, linked map, insertion order is relevant!
+  private LinkedHashMap<String, String> sorting = new LinkedHashMap<String, String>();
+  private LinkedList<String> closeStack = new LinkedList<String>();
+  private StringBuffer builder = new StringBuffer();
```
Comment (atennapel) [inline src/CommunityCommons/javasource/communitycommons/XPath.java:72]: As discussed in the standup, we may revert the commit that changes this StringBuilder to a StringBuffer.

File: src/CommunityCommons/javasource/communitycommons/XPath.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (lvanengelen) [reply src/CommunityCommons/javasource/communitycommons/XPath.java]: I updated the comment and synced the tokens with the ones on the documentation page. I included constants for `NULL` and `empty` as well.

File: src/CommunityCommons/javasource/communitycommons/XPath.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (lvanengelen) [reply src/CommunityCommons/javasource/communitycommons/XPath.java]: I removed superfluous `this` where I could find them. There were some lines like this:
```java
long count = this.count();
```
where I just renamed the count member variable to something else (even though `long count = count()` does the right thing).

File: src/CommunityCommons/javasource/communitycommons/XPath.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (lvanengelen) [reply src/CommunityCommons/javasource/communitycommons/XPath.java]: Done

File: src/CommunityCommons/javasource/communitycommons/XPath.java [hunk 10]
```diff
@@ -33,806 +33,829 @@
...
+      this.count = count;
+    }
+
+    @Override
+    public Boolean call() {
+      try {
+        batchProcessor.onItem(XPath.createProxy(Core.createSystemContext(), self.proxyClass, item), index, count); // mwe:
+                                                                                                                   // hmm,
+                                                                                                                   // many
+                                                                                                                   // contexts..
```
[the remark this answers is not in the document]
Comment (lvanengelen) [reply src/CommunityCommons/javasource/communitycommons/XPath.java]: Removed the comment, as I had no idea what it meant.

File: src/test/communitycommons/XPathTest.java (hunk 2 above)
Comment (lvanengelen) [reply src/test/communitycommons/XPathTest.java:53]: Added some tests

File: src/CommunityCommons/javasource/communitycommons/XPath.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (lvanengelen) [reply src/CommunityCommons/javasource/communitycommons/XPath.java]: I replaced the tabs with spaces but I kept the parentheses as they were. There are already other files with the same convention. The parentheses are also a less invasive change (i.e. not on every line) making it easier to diff.

File: src/CommunityCommons/javasource/communitycommons/XPath.java [hunk 11]
```diff
@@ -33,301 +33,372 @@
 import com.mendix.systemwideinterfaces.core.IMendixIdentifier;
 import com.mendix.systemwideinterfaces.core.IMendixObject;
 
-public class XPath<T>
-{
-	/** Build in tokens, see: https://world.mendix.com/display/refguide3/XPath+Keywords+and+System+Variables
-	 * 
+public class XPath<T> {
```
Comment (atennapel) [inline src/CommunityCommons/javasource/communitycommons/XPath.java:36]: The curly brace is on the same line here. That does not match with the generated actions. Should we still fix that?

**Pull request merged**